File: Sources/Domain/SystemRecoveryPolicy.swift
import Foundation

enum SystemEnvironmentBlocker: Hashable, Sendable {
  case sleep
  case inactiveSession
  case networkUnavailable
}

enum SystemRecoveryAction: Equatable, Sendable {
  case pause
  case resume
}

struct SystemRecoveryPolicy: Sendable {
  private var blockers: Set<SystemEnvironmentBlocker> = []

  var isAvailable: Bool {
    blockers.isEmpty
  }

  mutating func update(
    _ blocker: SystemEnvironmentBlocker,
    isBlocked: Bool
  ) -> SystemRecoveryAction? {
    let wasAvailable = isAvailable
    if isBlocked {
      blockers.insert(blocker)
    } else {
      blockers.remove(blocker)
    }

    guard wasAvailable != isAvailable else {
      return nil
    }
    return isAvailable ? .resume : .pause
  }
}

File: platform/windows/MihomoMeter.Windows.Core/Domain/SystemRecoveryPolicy.cs
namespace MihomoMeter.Windows.Core.Domain;

public enum SystemEnvironmentBlocker
{
    Sleep,
    InactiveSession,
    NetworkUnavailable,
}

public enum SystemRecoveryAction
{
    Pause,
    Resume,
}

public sealed class SystemRecoveryPolicy
{
    private readonly HashSet<SystemEnvironmentBlocker> _blockers = [];

    public bool IsAvailable => _blockers.Count == 0;

    public SystemRecoveryAction? Update(
        SystemEnvironmentBlocker blocker,
        bool isBlocked)
    {
        var wasAvailable = IsAvailable;
        if (isBlocked)
        {
            _blockers.Add(blocker);
        }
        else
        {
            _blockers.Remove(blocker);
        }

        if (wasAvailable == IsAvailable)
        {
            return null;
        }

        return IsAvailable
            ? SystemRecoveryAction.Resume
            : SystemRecoveryAction.Pause;
    }
}
